### PR TITLE
[#IR-98] : Added create API support for goal, highlight and notes

### PR DESCRIPTION
**Pull request by anmolghosh**

### CODE REVIEW

File: apps/retrospective/services/retrospective_feedback.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+		if retroFeedback.Type != models.GoalType {
+			return nil, errors.New("expectedAt can be updated only for goal " +
+				"type retrospective feedback")
+		}
+		retroFeedback.ExpectedAt = feedbackData.ExpectedAt
+	}
+
+	if feedbackData.AssigneeID != nil {
+		var userID uint
+		if err := db.Raw("SELECT user_teams.user_id FROM user_teams JOIN retrospectives "+
```
Comment (ishugupta92) [inline apps/retrospective/services/retrospective_feedback.go]: Can we change this to use GORM methods?

File: apps/retrospective/services/retrospective_feedback.go [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+			"ON retrospectives.team_id = user_teams.team_id WHERE retrospectives.id = ? "+
+			"and user_teams.user_id = ?;", retroID,
+			feedbackData.AssigneeID).Scan(&userID).Error; err != nil {
+			if err.Error() == "record not found" {
```
Comment (ishugupta92) [inline apps/retrospective/services/retrospective_feedback.go]: GORM provides a IsRecordNotFound method that we can use here

File: apps/retrospective/services/retrospective_feedback.go [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+				return nil, errors.New("cannot assign to requesxted user")
```
Comment (ishugupta92) [inline apps/retrospective/services/retrospective_feedback.go]: Typo in requested

File: apps/retrospective/services/retrospective_feedback.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return service.getRetrospectiveFeedback(retroFeedback.ID)
+
+}
+
+// Update ...
+func (service RetrospectiveFeedbackService) Update(userID uint, sprintID string, retroID string,
```
Comment (ishugupta92) [inline apps/retrospective/services/retrospective_feedback.go]: We aren't using sprintID in this method, so we can remove it.

File: apps/retrospective/services/retrospective_feedback.go [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+		Where("id = ?", sprintID).
+		Find(&sprint).Error; err != nil {
+		return nil, err
+	}
+
+	switch goalType {
+	case "added":
+		if err := db.Model(&models.RetrospectiveFeedback{}).
+			Where("retrospective_id = ? AND type = ?", retroID, models.GoalType).
+			Where("resolved_at IS NULL").
```
Comment (ishugupta92) [inline apps/retrospective/services/retrospective_feedback.go]: We extract most of this code outside the switch block

File: apps/retrospective/services/retrospective_feedback.go [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+	feedbackData *retrospectiveSerializers.RetrospectiveFeedbackCreateSerializer) (
+	*retrospectiveSerializers.RetrospectiveFeedback,
+	error) {
+	db := service.DB
+
+	retroIDInt, _ := strconv.Atoi(retroID)
+	sprint := models.Sprint{}
+
+	if err := db.Model(&models.Sprint{}).
+		Where("id = ?", sprintID).
```
Comment (ishugupta92) [inline apps/retrospective/services/retrospective_feedback.go]: We need to add a check that the Sprint is in `Active` state.

File: apps/retrospective/services/retrospective_feedback.go [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+			feedbackData.AssigneeID).Scan(&userID).Error; err != nil {
+			if err.Error() == "record not found" {
+				return nil, errors.New("cannot assign to requesxted user")
+			}
+			return nil, err
+		}
+		retroFeedback.AssigneeID = feedbackData.AssigneeID
+	}
+
+	err := db.Save(&retroFeedback).Error
```
Comment (ishugupta92) [inline apps/retrospective/services/retrospective_feedback.go]: We need a check that we cannot edit Notes and Highlights for a frozen sprint, and we cannot update a goal if it is resolved.

File: controllers/v1/sprint_goal.go [hunk 8]
```diff
@@ -28,68 +30,146 @@ func (ctrl SprintGoalController) Add(c *gin.Context) {
...
+		goalID,
+		true)
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{
+			"message": "Failed to resolved goal",
+			"error":   err.Error()})
+		return
+	}
+
+	c.JSON(http.StatusOK, response)
```
Comment (ishugupta92) [inline controllers/v1/sprint_goal.go:176]: Let's add an audit trail here

File: controllers/v1/sprint_goal.go [hunk 9]
```diff
@@ -28,68 +30,146 @@ func (ctrl SprintGoalController) Add(c *gin.Context) {
...
+		retroID,
+		goalID,
+		false)
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{
+			"message": "Failed to un-resolve goal",
+			"error":   err.Error()})
+		return
+	}
+
```
Comment (ishugupta92) [inline controllers/v1/sprint_goal.go:208]: Let's add an audit trail here

File: controllers/v1/sprint_highlight.go [hunk 10]
```diff
@@ -26,38 +28,90 @@ func (ctrl SprintHighlightController) Add(c *gin.Context) {
...
+		retroID,
+		highlightID,
+		&feedbackData)
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{
+			"message": "Failed to update highlight",
+			"error":   err.Error()})
+		return
+	}
+
```
Comment (ishugupta92) [inline controllers/v1/sprint_highlight.go:133]: Let's add an audit trail here and in other update actions

File: apps/retrospective/services/retrospective_feedback.go [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+	userServices "github.com/iReflect/reflect-app/apps/user/services"
+	"github.com/jinzhu/gorm"
+	"github.com/pkg/errors"
+	"strconv"
+)
+
+// RetrospectiveFeedbackService ...
+type RetrospectiveFeedbackService struct {
+	DB          *gorm.DB
+	TeamService userServices.TeamService
```
Comment (ishugupta92) [inline apps/retrospective/services/retrospective_feedback.go]: This isn't being used right now

File: apps/retrospective/services/retrospective_feedback.go [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+		}
+		retroFeedback.ExpectedAt = feedbackData.ExpectedAt
+	}
+
+	if feedbackData.AssigneeID != nil {
+		var userID uint
+		if err := db.Raw("SELECT user_teams.user_id FROM user_teams JOIN retrospectives "+
+			"ON retrospectives.team_id = user_teams.team_id WHERE retrospectives.id = ? "+
+			"and user_teams.user_id = ?;", retroID,
+			feedbackData.AssigneeID).Scan(&userID).Error; err != nil {
```
Comment (ishugupta92) [inline apps/retrospective/services/retrospective_feedback.go]: We cannot use a `uint` as a Scan target, it has to be a `struct` or a `slice`

File: apps/retrospective/serializers/retrospective_feedback.go [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+type RetrospectiveFeedbackUpdateSerializer struct {
+	Text       *string    `json:"Text"`
+	Scope      *int8      `json:"Scope"`
+	AssigneeID *uint      `json:"AssigneeID"`
+	ExpectedAt *time.Time `json:"ExpectedAt"`
+}
+
+// RetrospectiveFeedbackCreateSerializer ...
+type RetrospectiveFeedbackCreateSerializer struct {
+	SubType string `json:"subType" binding:"required"`
```
Comment (ishugupta92) [inline apps/retrospective/serializers/retrospective_feedback.go:36]: We need to add validations for enums

File: apps/retrospective/models/retrospective_feedback.go [hunk 14]
```diff
@@ -58,8 +58,8 @@ type RetrospectiveFeedback struct {
 	RetrospectiveID uint                       `gorm:"not null"`
 	Text            string                     `gorm:"type:text; not null"`
 	Scope           RetrospectiveFeedbackScope `gorm:"default:0; not null"`
-	AssigneeID      uint
-	Assignee        userModels.User
+	AssigneeID      *uint
+	Assignee        *userModels.User
```
Comment (ishugupta92) [inline apps/retrospective/models/retrospective_feedback.go]: We might need validations for addedDate < ResolvedAt, ExpectedAt

File: apps/retrospective/services/retrospective_feedback.go [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+
+// Add ...
+func (service RetrospectiveFeedbackService) Add(userID uint, sprintID string, retroID string,
+	feedbackType models.RetrospectiveFeedbackType,
+	feedbackData *retrospectiveSerializers.RetrospectiveFeedbackCreateSerializer) (
+	*retrospectiveSerializers.RetrospectiveFeedback,
+	error) {
+	db := service.DB
+
+	retroIDInt, _ := strconv.Atoi(retroID)
```
Comment (ishugupta92) [inline apps/retrospective/services/retrospective_feedback.go]: Let's handle the case where this fails

**Review by ishugupta92 — CHANGES_REQUESTED** [review]

File: apps/retrospective/models/retrospective_feedback.go (hunk 14 above)
Comment (anmolghosh) [reply apps/retrospective/models/retrospective_feedback.go]: done

File: apps/retrospective/serializers/retrospective_feedback.go (hunk 13 above)
Comment (anmolghosh) [reply apps/retrospective/serializers/retrospective_feedback.go:36]: Not needed for now as sub-type is not defined/hardcoded in string

File: apps/retrospective/services/retrospective_feedback.go (hunk 11 above)
Comment (anmolghosh) [reply apps/retrospective/services/retrospective_feedback.go]: done

File: apps/retrospective/services/retrospective_feedback.go (hunk 15 above)
Comment (anmolghosh) [reply apps/retrospective/services/retrospective_feedback.go]: done

File: apps/retrospective/serializers/retrospective_feedback.go (hunk 13 above)
Comment (ishugupta92) [reply apps/retrospective/serializers/retrospective_feedback.go:36]: We will still need it for Scope

File: apps/retrospective/services/retrospective_feedback.go (hunk 6 above)
Comment (anmolghosh) [reply apps/retrospective/services/retrospective_feedback.go]: done

File: apps/retrospective/services/retrospective_feedback.go (hunk 4 above)
Comment (anmolghosh) [reply apps/retrospective/services/retrospective_feedback.go]: done

File: apps/retrospective/services/retrospective_feedback.go (hunk 1 above)
Comment (anmolghosh) [reply apps/retrospective/services/retrospective_feedback.go]: we can ignore here

File: apps/retrospective/services/retrospective_feedback.go (hunk 12 above)
Comment (anmolghosh) [reply apps/retrospective/services/retrospective_feedback.go]: replaced with count, not consuming the value here

File: apps/retrospective/services/retrospective_feedback.go (hunk 2 above)
Comment (anmolghosh) [reply apps/retrospective/services/retrospective_feedback.go]: ignore for now

File: apps/retrospective/services/retrospective_feedback.go (hunk 12 above)
Comment (anmolghosh) [reply apps/retrospective/services/retrospective_feedback.go]: fixed

File: apps/retrospective/services/retrospective_feedback.go (hunk 3 above)
Comment (anmolghosh) [reply apps/retrospective/services/retrospective_feedback.go]: fixed

File: apps/retrospective/services/retrospective_feedback.go (hunk 7 above)
Comment (anmolghosh) [reply apps/retrospective/services/retrospective_feedback.go]: The first check is handled by the  `UserCanEditSprint` method.
Added code for second

File: apps/retrospective/services/retrospective_feedback.go (hunk 5 above)
Comment (anmolghosh) [reply apps/retrospective/services/retrospective_feedback.go]: done

File: controllers/v1/sprint_goal.go (hunk 8 above)
Comment (anmolghosh) [reply controllers/v1/sprint_goal.go:176]: done

File: controllers/v1/sprint_goal.go (hunk 9 above)
Comment (anmolghosh) [reply controllers/v1/sprint_goal.go:208]: done

File: controllers/v1/sprint_highlight.go (hunk 10 above)
Comment (anmolghosh) [reply controllers/v1/sprint_highlight.go:133]: done

File: apps/retrospective/services/retrospective_feedback.go (hunk 3 above)
Comment (anmolghosh) [reply apps/retrospective/services/retrospective_feedback.go]: done

File: apps/retrospective/services/retrospective_feedback.go (hunk 5 above)
Comment (anmolghosh) [reply apps/retrospective/services/retrospective_feedback.go]: done

File: apps/retrospective/serializers/retrospective_feedback.go (hunk 13 above)
Comment (anmolghosh) [reply apps/retrospective/serializers/retrospective_feedback.go:36]: done

**Review by ishugupta92 — APPROVED** [review]

**Pull request merged**